[Lens] Don't block render on missing field

drewdaemon opened the pull request:
## Summary

Resolve https://github.com/elastic/kibana/issues/143673
Resolve https://github.com/elastic/kibana/issues/147485

This PR introduces the new error/warning UI outlined in https://github.com/elastic/kibana/issues/147485. In addition
1. Missing fields are now treated as non-blocking errors (https://github.com/elastic/kibana/issues/143673) and are shown in the toolbar, the dimension button, and the embeddable 
2. Precision warnings are now shown on the corresponding dimension button as well as in the editor toolbar. I included this so that there is an example of warnings on dimension buttons to see.

### Screenshots
#### Editor

##### Errors
<img width="332" alt="Screen Shot 2023-01-26 at 11 43 58 AM" src="https://user-images.githubusercontent.com/315764/214910074-8e66b9ae-8d3b-4232-b463-f68da8c5be09.png">

<img width="508" alt="Screen Shot 2023-01-26 at 11 43 41 AM" src="https://user-images.githubusercontent.com/315764/214910076-f9d88f03-98c1-4405-82a3-8e930c737d27.png">

##### Warnings
<img width="352" alt="Screen Shot 2023-01-26 at 11 44 14 AM" src="https://user-images.githubusercontent.com/315764/214910069-9ebc8620-fe67-48cd-9640-cc3824a3e62e.png">
<img width="546" alt="Screen Shot 2023-01-26 at 12 14 26 PM" src="https://user-images.githubusercontent.com/315764/214916696-f194e31c-3f41-47ad-a40e-c1f184cce5bc.png">

##### Both
<img width="365" alt="Screen Shot 2023-01-26 at 9 17 05 AM" src="https://user-images.githubusercontent.com/315764/214910152-e141173d-4966-4552-b42c-bae2eb2bf50e.png">

#### Embeddable

##### Errors
<img width="354" alt="Screen Shot 2023-01-27 at 9 26 11 AM" src="https://user-images.githubusercontent.com/315764/215123497-55da1a1f-5d1e-4398-853b-79d8dcf0bd97.png">


##### Warnings
<img width="547" alt="Screen Shot 2023-01-27 at 9 21 30 AM" src="https://user-images.githubusercontent.com/315764/215123522-8ecd7c6a-53a1-4051-9b5f-0ec00fabe963.png">


##### Both
<img width="382" alt="Screen Shot 2023-01-27 at 9 25 17 AM" src="https://user-images.githubusercontent.com/315764/215123568-110ee740-c96f-4dab-9624-90e5f5476aad.png">

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios (flaky runner [in progress](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/1809)).
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))

andrewctate commented:
@MichaelMarcialis I've posted screenshots of how this looks right now. As you can see, the error is showing up in the correct locations and the render is no longer being blocked.

We can move forward once you've supplied feedback including the desired iconography.

It would be awesome to get this in by feature freeze (Feb 7) if you have the room on your plate.

MichaelMarcialis commented:
> @MichaelMarcialis I've posted screenshots of how this looks right now. As you can see, the error is showing up in the correct locations and the render is no longer being blocked.
> 
> We can move forward once you've supplied feedback including the desired iconography.
> 
> It would be awesome to get this in by feature freeze (Feb 7) if you have the room on your plate.

Thanks, @andrewctate. I'm putting together some quick mockups based on your screenshots above and our previous conversations. Hope to get them to you ASAP. Stay tuned.

MichaelMarcialis commented:
Hey, @andrewctate. Per our chat today, I've put those [quick mockups](https://www.figma.com/file/LAaZdPNqcJjOsWt8KO7104/Mockups-%2F-Version-1-%5BIn-Progress%5D?node-id=0%3A1&t=bgnd3xuRHcZanqmE-1) in a Figma file for you. I'll also send you SVGs for the error and warning icons for you to use as custom icons, in case they don't update EUI in Kibana before feature freeze. Nonetheless, I'll CC you a PR I'll be cutting shortly to add/update those icons in EUI.

drewdaemon commented:
@KOTungseth 

I'm hoping to get feedback on the wording and length of the following warning messages which are displayed in tooltips on the dimension buttons. They match longer descriptions including doc links and quick actions that are shown in the menu on the toolbar.

> This might be an approximation. You can enable accuracy mode for more precise results, but note that it increases the load on the Elasticsearch cluster.

This is shown when they have a top values function configured, we're getting feedback from Elasticsearch that the results may be approximate, and accuracy mode disabled in the dimension editor's advanced settings (originally added in https://github.com/elastic/kibana/pull/129220).

> This might be an approximation. For more precise results, try increasing the number of top values or using filters instead.

This is shown when they have a top values function configured, we're getting feedback from Elasticsearch that the results may be approximate, but they've already enabled accuracy mode (originally added in https://github.com/elastic/kibana/pull/129220).

> This may be approximate due to how the data is indexed. Try sorting by rarity instead of ascending count of records.

This is shown when they have a top values function configured, they are sorting by an ascending count, the field isn't floating point type, and they aren't using multiple fields in their top values function (original PR: https://github.com/elastic/kibana/pull/121500)

drewdaemon commented on x-pack/plugins/lens/public/datasources/text_based/text_based_languages.tsx:
```diff
@@ -233,13 +233,14 @@ export function getTextBasedDatasource({
-      if (!column || !indexPattern) return false;
-      return true;
-    },
+    // TODO - make sure this is covered in the user messsages
+    // isValidColumn(state, indexPatterns, layerId, columnId) {
+    //   const layer = state.layers[layerId];
+    //   const column = layer.columns.find((c) => c.columnId === columnId);
+    //   const indexPattern = indexPatterns[layer.index];
+    //   if (!column || !indexPattern) return false;
+    //   return true;
+    // },
```
@stratoula I'd like your input here.

This method performed a column existence check and an index pattern existence check. The result was used to mark the dimension button invalid.

To me, it looks like we may be able to delete this without replacement.

The column check is covered in the [`renderDimensionTrigger`](https://github.com/elastic/kibana/blob/da0b63b3e56a318e61f1a6ce36723bd4a07c9b76/x-pack/plugins/lens/public/datasources/text_based/text_based_languages.tsx#L388) method and I've verified that the UI looks the same.

The last thing I'm wondering: do you know why the index pattern check is in here? E.g. is there a way to hit this branch?

stratoula replied:
Yes @drewdaemon we can clear this out

drewdaemon commented on x-pack/plugins/lens/public/datasources/form_based/time_shift_utils.tsx:
```diff
@@ -171,6 +172,35 @@ export function getLayerTimeShiftChecks({
   };
 }
 
+export function getDisallowedPreviousShiftMessage(
```
TODO — remove. This was removed on main but got added back here during a merge.

drewdaemon commented on x-pack/plugins/lens/public/state_management/types.ts:
```diff
@@ -38,7 +38,6 @@ export interface PreviewState {
   visualization: VisualizationState;
   datasourceStates: DatasourceStates;
   activeData?: TableInspectorAdapter;
-  requestWarnings?: string[];
```
use of these was removed in https://github.com/elastic/kibana/pull/147818, but this was left over in the types

elasticmachine commented:
Pinging @elastic/kibana-visualizations @elastic/kibana-visualizations-external (Team:Visualizations)

drewdaemon commented:
@elasticmachine merge upstream

stratoula commented:
@drewdaemon I have this configuration. As long as the terms agg is optional dimension shouldn't this been rendered? Show only the horizontal and the vertical axis as the fields exist and not the breakdown.

<img width="2496" alt="image" src="https://user-images.githubusercontent.com/17003240/215458679-57289c70-1248-4906-bf7d-5d72e994d34f.png">

KOTungseth commented:
> > ...might be an approximation. You can enable accuracy mode for more precise results, but note that it increases the load on the Elasticsearch cluster.

How about:

`...might be an approximation. For more precise results, you can enable accuracy mode, but it increases the load on the Elasticsearch cluster.`



> > ...might be an approximation. For more precise results, try increasing the number of top values or using filters instead.

How about:

`...might be an approximation. For more precise results, increase the number of top values or apply filters.`



> > ...may be approximate due to how the data is indexed. Try sorting by rarity instead of ascending count of records.

How about:

`...may be approximate depending on how the data is indexed. For more precise results, sort by rarity.`

cc: @gchaps

drewdaemon commented:
@stratoula 

>As long as the terms agg is optional dimension shouldn't this been rendered? Show only the horizontal and the vertical axis as the fields exist and not the breakdown. (but still see the error on the dimension level)

This is a good find.

Looking at the tables and the responses for each case, it looks like the tabify logic in the data plugin generates an empty table when the terms agg's buckets are missing. That's why the no-data shows up in that case.

#### With breakdown on missing field

(terms agg ID is `1` here; since buckets is empty, no data)
<img width="425" alt="Screenshot 2023-01-30 at 11 47 58 AM" src="https://user-images.githubusercontent.com/315764/215556278-15f42f9d-96f1-4f6c-995c-c47fc9de5cd2.png">

<img width="488" alt="Screenshot 2023-01-30 at 11 50 34 AM" src="https://user-images.githubusercontent.com/315764/215556358-d922b97c-cffb-403f-9190-75b837cde7bf.png">

#### With no breakdown
<img width="418" alt="Screenshot 2023-01-30 at 11 48 07 AM" src="https://user-images.githubusercontent.com/315764/215556586-f172ef09-2aaa-4651-80e8-d462ba3ba7b6.png">

<img width="242" alt="Screenshot 2023-01-30 at 11 50 40 AM" src="https://user-images.githubusercontent.com/315764/215556475-806997f6-5545-428a-95fc-6e93fb4c7661.png">

I think the best way to fix this would be to trim the top values dimension from the Lens config before generating the expression. However, I left this out of scope: https://github.com/elastic/kibana/issues/144906. I could always look into the `tabify` code, but it feels like overkill 😅 

How big an impact do you think this disparity has on the UX (cc: @MichaelMarcialis )? I think we can resolve it, but it is naturally part of a bigger effort.

MichaelMarcialis commented:
> @MichaelMarcialis how do you feel about this default icon placement on smaller screens?

If you mean the repositioning of the `alert` icon to be above the "Top 50…" text, I prefer it with the icon to always appear to the left of the text, even at small viewport sizes. If you're using the `EuiFlexGroup` component, you should be able to achieve that by adding a `responsive={false}` prop.

gchaps commented:
@KOTungseth's comments on the text look.  Here are a few more comments:

> 1 of 2 shards failed
>The data you are seeing might be wrong or incomplete.

We don't use "see" for accessibility reasons.  How about:

`The data might be wrong or incomplete.`

---

> Field fun was not found.

Can you put "fun" in a code font?

Field `fun` was not found.

Also, our guideline is to use an ending period in tooltip text. However, it's more important to be consistent in the app, so if other tooltips don't use an ending period, skip.

---

> Top 27 values of clients…..

The "Learn more" link isn’t blue.  Is the link needed here?

---

> Field to delete was not found.

"Field to delete" sounds a little awkward.  Instead, can you use the same form as the tooltip above:

Field `fun` was not found.

drewdaemon commented:
Thanks @gchaps 

> We don't use "see" for accessibility reasons. How about:

Updated.

> Can you put "fun" in a code font?

> Also, our guideline is to use an ending period in tooltip text. However, it's more important to be consistent in the app, so if other tooltips don't use an ending period, skip.

Done

<img width="576" alt="Screenshot 2023-01-31 at 5 18 25 PM" src="https://user-images.githubusercontent.com/315764/215905858-3b451f8c-4bf7-4635-859f-6d1ed847f078.png">

> The "Learn more" link isn’t blue. Is the link needed here?

Removed

>"Field to delete" sounds a little awkward. Instead, can you use the same form as the tooltip above:

This is the case. "to delete" was the field name. Things like this will be much more clear now that I implemented your code formatting suggestion!

stratoula reviewed (approved):
This works great Drew! Code LGTM. I tested with various scenarios and works as expected. It would be awesome to have the same behavior when a terms field is missing but as a first iteration is great!

drewdaemon commented:
@elasticmachine merge upstream

lukasolson reviewed (approved):
Search changes LGTM

MichaelMarcialis commented:
> > Can you put "fun" in a code font?
> 
> > Also, our guideline is to use an ending period in tooltip text. However, it's more important to be consistent in the app, so if other tooltips don't use an ending period, skip.
> 
> Done
> 
> <img alt="Screenshot 2023-01-31 at 5 18 25 PM" width="576" src="https://user-images.githubusercontent.com/315764/215905858-3b451f8c-4bf7-4635-859f-6d1ed847f078.png">

@gchaps: If possible, could we avoid using the `EuiCode` component for styling the field names in warnings and errors? Using it raises the following concerns:

- The `EuiCode` component wraps text in a `code` element, which wouldn't be semantically correct in this instance (potential accessibility issue).
- The addition of the purple color doesn't look great to me, especially when we're attempting to use color to quickly identify warning (yellow) versus error (red). 
- The `EuiCode` component wasn't created with long strings of text in mind (i.e. text that requires the ability to break a line). As field names can become quite long, I think it would be best avoided. 
- We don't use `EuiCode` for field names anywhere else in Lens, so it's introduces some level of inconsistency.

Thoughts? If you are in agreement, I suggest we just use simple bolding for field names. CCing @drewdaemon.

MichaelMarcialis reviewed (commented):
Thanks so much for putting this together, @drewdaemon! Very well done. In addition to my previous comment on the use of `EuiCode`, I'm leaving a few comments below for my initial review.

Also, I may have found a small bug. When using the `clientip` field from the `Kibana Sample Data Logs` data view to produce the `Top 3 values of clientip might be an approximation.` warning, I noticed that the warning wasn't showing up when the visualization was placed in a dashboard. Is that intentional?

![image](https://user-images.githubusercontent.com/3884767/216375177-002bd2f2-a2d3-48f5-bc87-b19a65daa169.png)

![image](https://user-images.githubusercontent.com/3884767/216375339-6b06bfcb-e38c-437b-a090-4a18bcb5f734.png)

drewdaemon commented:
@MichaelMarcialis thanks for the feedback. Happy to make the requested adjustments.

> the EuiCode component wraps text in a code element, which wouldn't be semantically correct in this instance (potential accessibility issue).

I verified that the tooltip text was accessible to a screenreader, however the rest of your reasons for preferring `bold` typography make sense.

> When using the clientip field from the Kibana Sample Data Logs data view to produce the Top 3 values of clientip might be an approximation. warning, I noticed that the warning wasn't showing up when the visualization was placed in a dashboard. Is that intentional?

This warning wasn't shown on the dashboard before and I didn't change that. We can decide to show it, but just be aware that it's a change in behavior. WDYT?

MichaelMarcialis commented:
@drewdaemon: Thanks for those comments. Some quick responses below:

> > the EuiCode component wraps text in a code element, which wouldn't be semantically correct in this instance (potential accessibility issue).
> 
> I verified that the tooltip text was accessible to a screenreader, however the rest of your reasons for preferring `bold` typography make sense.

Thanks for checking! In this case though, my accessibility concern was more about the semantics of the element and that the string would be presented as a chunk of code to users, when it's actually not code.

<hr/>

> > When using the clientip field from the Kibana Sample Data Logs data view to produce the Top 3 values of clientip might be an approximation. warning, I noticed that the warning wasn't showing up when the visualization was placed in a dashboard. Is that intentional?
> 
> This warning wasn't shown on the dashboard before and I didn't change that. We can decide to show it, but just be aware that it's a change in behavior. WDYT?

Interesting. Ultimately, I'll defer to the product folks on this. To my mind though, unless the warning doesn't make sense in the context of a dashboard or embed, I think it would make sense to include it (minus the corrective action button, since the user is outside the editor at that point). CCing @ninoslavmiskovic, in case he has an opinion on it.

<hr/>

I had one additional last-minute thought. Originally, I was keen on the notion of having the error/warning icons replace any other prepended icons that were shown on the related dimension item (color swatch, swatch with slash, collapse, hidden, etc.), as you've shown here in your PR. Upon further reflection and after testing, I'm now thinking it may be too heavy-handed for us to replace those icons outright (as some meta information about the dimension may be lost in the process in some instances. Instead, could we show both, but have it so the error/warning icons appear adjacent to the text (rather than sandwiching the other icon in between), as they share the same color? Example below:

![image](https://user-images.githubusercontent.com/3884767/216398096-7811c258-ab5d-448c-b9db-e0249dfe54e1.png)

I don't think this is critical though, so if you'd prefer this to be spun off as a separate issue, let me know.

drewdaemon commented:
@MichaelMarcialis 

> Instead, could we show both, but have it so the error/warning icons appear adjacent to the text (rather than sandwiching the other icon in between), as they share the same color?

Done. This is what it looks like

<img width="423" alt="Screenshot 2023-02-02 at 12 04 50 PM" src="https://user-images.githubusercontent.com/315764/216407847-a8d9afea-0c93-4b09-a0ea-97f361adfaa3.png">


WDYT?

drewdaemon commented:
@MichaelMarcialis also, here's the new dimension button styling with a color palette dimension

<img width="423" alt="Screenshot 2023-02-02 at 11 49 24 AM" src="https://user-images.githubusercontent.com/315764/216408976-46751c29-90bb-469c-86e9-5553c03c8880.png">

When the delete button is hovered, it runs into the color palette indicator. I'll leave it to you to decide whether that's an issue.
<img width="399" alt="Screenshot 2023-02-02 at 12 10 02 PM" src="https://user-images.githubusercontent.com/315764/216409218-438bd361-c147-4ad6-84c4-7fb8982818a9.png">

ninoslavmiskovic commented:
A couple of comments @MichaelMarcialis  and @drewdaemon :

- I think it is important to showcase to the user that he/she should focus on the warning, and therefore it make sense to overwrite the existing icon to nudge the user to hover over the warning and figure out what is wrong. 
- Adding to icons does not feels scalable - What if we want to add an "i" icon in the future for information etc. 

I have in similar indicator exercises gotten lucky and used text and the field border to indicate something is wrong. Red text + red border. Or even making the background of the form field red.

drewdaemon commented:
@MichaelMarcialis here are the icons with the new padding.


<img width="611" alt="Screenshot 2023-02-02 at 2 14 49 PM" src="https://user-images.githubusercontent.com/315764/216440727-849d68c1-1774-4a97-8332-0f2bda072d49.png">
<img width="197" alt="Screenshot 2023-02-02 at 2 13 42 PM" src="https://user-images.githubusercontent.com/315764/216440729-5e9a7d23-6d36-49df-8bd7-33a8c427fcce.png">


@ninoslavmiskovic thanks for your feedback. I believe @MichaelMarcialis will be out for the first of next week, so if you don't mind, I think we can take your objections to the dimension button icon pattern as a follow up discussion. It's easily changed—if we decide to go another route it can make it in for 8.8 or even sneak in as a bug fix after feature freeze for 8.7 if need be.

With feature freeze looming, I don't want to hold up this PR.

MichaelMarcialis commented:
> A couple of comments @MichaelMarcialis and @drewdaemon :
> 
> * I think it is important to showcase to the user that he/she should focus on the warning, and therefore it make sense to overwrite the existing icon to nudge the user to hover over the warning and figure out what is wrong.
> * Adding to icons does not feels scalable - What if we want to add an "i" icon in the future for information etc.
> 
> I have in similar indicator exercises gotten lucky and used text and the field border to indicate something is wrong. Red text + red border. Or even making the background of the form field red.

@ninoslavmiskovic: For dimension items specifically, I would try to avoid altering the border or background color in an error or warning state (at least, not without some other stylistic changes first). The border styling is currently being used to on hover to indicate drag-and-drop. The changing the background color also becomes problematic/conflicting when you take into account the other elements of color already being used in the dimension items (color swatches, color palettes, delete button, etc.). Not saying we can't do these things, but we'd need to make other adjustments to accommodate.

As for retaining the swatch/visibility/collapse icon before the error/warning icon, I'll ultimately defer to you on that. I agree with your scalability concerns. I was worried that the user may lose some context about the dimension in the process, but if you feel that loss is not significant or important, we can just have the error/warning icons overtake those other icon prepends as @drewdaemon previously had it. As he mentioned, I'm out Monday and Tuesday next week, so I'm good with whatever ya'll decide there.

MichaelMarcialis commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/config_panel/buttons/dimension_button.tsx:
```diff
@@ -25,33 +27,47 @@ export function DimensionButton({
-        iconType="cross"
-        iconSize="s"
+        iconType="trash"
         size="s"
```
Possible to reduce the size here to `xs` to fit more comfortably in the dimension item box?

```suggestion
        size="xs"
```

MichaelMarcialis commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/config_panel/buttons/dimension_button.tsx:
```diff
@@ -25,33 +27,47 @@ export function DimensionButton({
-      >
-        <ColorIndicator accessorConfig={accessorConfig}>{children}</ColorIndicator>
-      </EuiLink>
+      <EuiFlexItem>
```
Is `EuiFlexItem` needed here? May be able to replace with a regular old `div`.

MichaelMarcialis commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/workspace_panel/message_list.tsx:
```diff
@@ -0,0 +1,133 @@
+            title={buttonLabel}
+            css={customButtonStyles}
+          >
+            {errorCount > 0 && (
+              <>
+                <EuiIcon type={IconError} />
+                {errorCount}
+              </>
+            )}
+            {warningCount > 0 && (
+              <>
+                <EuiIcon
+                  type={IconWarning}
+                  css={css`
+                    margin-left: 4px;
+                  `}
+                />
+                {warningCount}
+              </>
+            )}
```
Would it be possible to give the error and warning icons a prop of `size="s"` when used on dashboard panels and embeds (but not in the Lens toolbar) to better match the designs?

MichaelMarcialis reviewed (approved):
Thanks for making those changes, @drewdaemon! I've left some small comments below for my re-review, but nothing worth blocking you on. I'll go ahead and approve now, as I'll be out early next week. Great work here.

As for the outstanding questions on how to handle the prepended icons and whether to show that warning I mentioned previous on dashboards/embeds, I'll defer to whatever you and the team decide.

drewdaemon commented:
@MichaelMarcialis I've merged the PR with the understanding that I will address your final round of feedback in https://github.com/elastic/kibana/issues/150241
